### DISCUSSION TITLE
Importing Data

### DISCUSSION THREAD

**Discussion by Obalus2022** [casact/chainladder-python#316] [Q&A] [answered]
Hey, just want to start by saying this is an exciting package.  after reserving in excel for years, I am really excited for find a package that handles the data in this way and creates an environment that follows the way that I think about data and process.

I am simultaneously working my way up both the Python and the chainladder learning curve, so forgive a noob question

I am trying to pull in quarterly development triangles where the format looks like this YYYY-Qq, so for example 2021Q4.  chain ladder seems to take the data in, although it does throw a warning about data being out of bounds, but when I try to look at the data, it is giving me garbage.  see example.  Pulling from Access

![image](https://user-images.githubusercontent.com/110035421/181005561-a7502c12-c808-4786-8c2d-87ea827b0cbc.png)

![image](https://user-images.githubusercontent.com/110035421/181005977-dc755806-5684-45d9-b212-6c88b6ec56bf.png)

**Comment by jbogaardt** [comment]
Hi @Obalus2022 , 

Syntactically, your code looks fine. The warning about some items having some development dates before the origin date indicates the negative development ages, so the data should be checked. 

Some other possible things to check:
1. Ensure the columns used for origin/development are indeed the appropriate columns for the triangle you want to create.
2. If the data you're importing is incremental, then `cumulative=False` should be used when creating a triangle.

Hope this helps, if not, I may need you to provide a bit more context and/or a reproducible example with fake data.

**Reply by Obalus2022** [reply]
So this is what a simplified version of my data looks like:

AccidentQuarter | ReportedQuarter | RiskCode | PDLALAE | OSLR

2006-Q1 | 2006-Q1 | ARP | 1199961.6 | 0.0
2006-Q1 | 2006-Q2 | ARP | 0.0 | 0.0
2006-Q2 | 2006-Q2 | ARP | 106595.2 | 0.0
2006-Q2 | 2006-Q3 | ARP | 481599.2 | 0.0
2006-Q3 | 2006-Q3 | ARP | 1116264.8 | 0.0
2006-Q3 | 2006-Q4 | ARP | 217327.6 | 0.0
2006-Q3 | 2007-Q1 | ARP | 211620.0 | 0.0
2006-Q3 | 2007-Q2 | ARP | 0.0 | 0.0
2006-Q4 | 2006-Q4 | ARP | 358700.0 | 0.0
2006-Q4 | 2007-Q1 | ARP | 56620.0 | 0.0
2006-Q4 | 2007-Q2 | ARP | 1213920.4 | 0.0
2006-Q4 | 2008-Q1 | ARP | 908103.6 | 0.0
2006-Q4 | 2008-Q4 | ARP | 0.0 | 0.0

and this is what Triangles is putting out.

![image](https://user-images.githubusercontent.com/110035421/181246252-7579d1a0-6afe-4548-9f9c-8d176b031681.png)

**Comment by jbogaardt** [comment]
Does this look like what you want?
```python
import chainladder as cl
import pandas as pd
from io import StringIO



data = """
AccidentQuarter|ReportedQuarter|RiskCode|PDLALAE|OSLR
2006-Q1|2006-Q1|ARP|1199961.6| 0.0
2006-Q1|2006-Q2|ARP|0.0| 0.0
2006-Q2|2006-Q2|ARP|106595.2| 0.0
2006-Q2|2006-Q3|ARP|481599.2| 0.0
2006-Q3|2006-Q3|ARP|1116264.8| 0.0
2006-Q3|2006-Q4|ARP|217327.6| 0.0
2006-Q3|2007-Q1|ARP|211620.0| 0.0
2006-Q3|2007-Q2|ARP|0.0| 0.0
2006-Q4|2006-Q4|ARP|358700.0| 0.0
2006-Q4|2007-Q1|ARP|56620.0| 0.0
2006-Q4|2007-Q2|ARP|1213920.4| 0.0
2006-Q4|2008-Q1|ARP|908103.6| 0.0
2006-Q4|2008-Q4|ARP|0.0| 0.0"""

triangle = cl.Triangle(
    pd.read_csv(StringIO(data), delimiter='|'),
    origin='AccidentQuarter', development='ReportedQuarter',
    index='RiskCode', columns=['PDLALAE', 'OSLR'],
    origin_format='%Y-%q', development_format='%Y-%q',
    cumulative=False)

triangle[triangle['RiskCode']=='ARP']['PDLALAE'].sum(axis='index').incr_to_cum()
```

**Reply by Obalus2022** [reply]
I think your assuming that data is a text string,  Its actually a dataframe.

![image](https://user-images.githubusercontent.com/110035421/181381775-b45818b5-c4b0-4770-9c6f-e5d770542d87.png)

**Comment by jbogaardt** [comment]
Ok, while the data string is only part of a larger dataframe, I've used it to create a reproducible example. 

Even though that text is not complete, the resultant triangle looks as I'd expect given I don't have all the data:
![image](https://user-images.githubusercontent.com/19401133/181386108-ecec4c44-0a28-4123-bcb8-af441f6032dd.png)

Do you expect something different?

**Reply by Obalus2022** [reply]
Im a little confused as I see that your method works, but not sure what to do about mine.  Are you suggesting that I turn my data frame into a string and then then read it in as a string. That seems to be counterintuitive.  Oddly I did try that, but I still got the same result. Could the problem be with my origin and development formats?

![image](https://user-images.githubusercontent.com/110035421/181396037-6b9c0900-b2a6-4500-b930-9723f75cb354.png)

**Comment by jbogaardt** [comment, marked as answer]
No, I do not suggest you convert your data to a string. Let's look instead at the difference in the code (chainladder syntax only) with expanded commentary:
```python
(
    triangle[triangle['RiskCode']=='ARP']['PDLALAE'] # Filters for some index and column, Fine
     .sum(axis='index') # Deliberately summing the index axis. If there is only one index .e.g. ARP, this is unnecessary.
     .incr_to_cum() # This method takes our incremental triangle and makes it cumulative which is what I think you want.
)
```

Your original code with commentary:
```python
(
    triangle[triangle['RiskCode']=='ARP']['PDLALAE'] # Filters for some index and column, Fine
        .sum() # No axis is specified in sum, so this sums the origin axis since we have a (1,1,12,12) shaped triangle. origin is the first axis with length > 1
        .to_frame() # this converts it to a pandas dataframe which precludes using the estimators in chainladder.
)
```

It sounds to me like you just need to accumulate your triangle with `incr_to_cum` and avoid `to_frame` unless you're leaving the world of `chainladder-python`.

**Reply by Obalus2022** [reply]
Thanks so much appreciate all the help.

**Reply by Obalus2022** [reply]
So, I have a followup on this,  I have been able to get all of my loss data into Triangle and now I am attempting to do the same with premiums which come from a different database.  I was able to then get my premium database all the way into Triangles,  however when I tried to concat my premium and loss data frames into one dataframe, Triangle didnt like it: Am I doing something wrong?

![image](https://user-images.githubusercontent.com/110035421/182031018-17682d46-961c-488d-a360-6db7cfcaf042.png)

![image](https://user-images.githubusercontent.com/110035421/182031042-03e76fc7-12dc-4c2f-8654-6ca10d7f543c.png)

![image](https://user-images.githubusercontent.com/110035421/182031066-22900f9f-a30e-4cb9-85ef-11255c121d41.png)

![image](https://user-images.githubusercontent.com/110035421/182031101-ec9cee22-33d8-4eb9-b071-8ba26c0c0782.png)

![image](https://user-images.githubusercontent.com/110035421/182031215-aaf75f8b-c5be-4ba9-84a2-fd2b81e82e51.png)

![image](https://user-images.githubusercontent.com/110035421/182031250-dbf807c8-1b66-4c32-b596-5b1a2f739258.png)

Is using concat the best way to merge premium and loss data frames, or is there a better way?

**Reply by jbogaardt** [reply]
> Is using concat the best way to merge premium and loss data frames, or is there a better way?

This is exactly what I do when making policy/claim level triangles.

There is nothing apparent from the code you show that anything is wrong. In these cases, it is important that you have a reproducible example of the issue. Otherwise it becomes way too time-consuming and often times impossible to trouble shoot.

For example, here is an attempt at replicating this workflow with publicly available data:
```python
import chainladder as cl
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/casact/chainladder-python/master/chainladder/utils/data/clrd.csv')

loss = df[['GRCODE', 'LOB', 'AccidentYear', 'DevelopmentYear', 'CumPaidLoss']]
prem = df[['GRCODE', 'LOB', 'AccidentYear', 'DevelopmentYear', 'EarnedPremDIR']]

comb = pd.concat((loss, prem), axis=0)

cl.Triangle(
    comb, origin='AccidentYear', development='DevelopmentYear', 
    columns=['CumPaidLoss', 'EarnedPremDIR'], index=['GRCODE', 'LOB'], 
    cumulative=True)
```
Unfortunately, I cannot replicate your issue with the above even after this attempt.
